feature(actions): refresh tokens without booting the engine

hypeJunction opened the pull request:
Introduces /refresh-token handler that refreshes tokens without booting engine.

hypeJunction committed:
feature(actions): refresh tokens without booting the engine

Introduces /refresh-token handler that refreshes tokens without booting engine.

hypeJunction commented:
I am doing some profiling and I have noticed that requests to `/refresh_token` consume over 2Mb. This gets it down to ~250Kb. We could reduce it to under ~150Kb, if we refactored some of the functions to OO, e.g. `sanitize_string()` etc.

hypeJunction commented:
Alternatively, we can substitute `$session_id` with another value (ideas?), and use cookie value if session is not started.

hypeJunction commented on engine/tests/phpunit/bootstrap.php:
```diff
@@ -41,6 +41,7 @@ function _elgg_testing_application(\Elgg\Application $app = null) {
 	'simplecache_enabled' => false,
 	'system_cache_enabled' => false,
 	'Elgg\Application_phpunit' => true,
+	'site_config_loaded' => true,
```
`\Elgg\Config` is trying to connect to the DB without this.

hypeJunction commented:
Ignore the numbers I posted earlier. They make no sense. I was looking at the wrong thing. 
This cuts off a few hundred milliseconds from the response time.

mrclay commented on engine/classes/Elgg/Application.php:
```diff
@@ -408,6 +408,13 @@ public function run() {
 
+		if (0 === strpos($path, '/refresh-token')) {
+			$this->loadCore();
+			$this->services->session->start();
```
It's fine for now, but I feel like this shouldn't be necessary. Like ElggSession get()/set() should make sure of that.

mrclay replied:
Looking into Symfony's depths, I think get/set already auto-starts.

hypeJunction replied:
I think our implementation reads from the database.
On Apr 13, 2016 4:21 PM, "Steve Clay" <notifications@github.com> wrote:

> In engine/classes/Elgg/Application.php
> <https://github.com/Elgg/Elgg/pull/9675#discussion_r59556136>:
>
> > @@ -408,6 +408,13 @@ public function run() {
> >  			return true;
> >  		}
> >
> > +		if (0 === strpos($path, '/refresh-token')) {
> > +			$this->loadCore();
> > +			$this->services->session->start();
>
> It's fine for now, but I feel like this shouldn't be necessary. Like
> ElggSession get()/set() should make sure of that.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/Elgg/Elgg/pull/9675/files/cc9cccf1ef891e8b9fcf3e04c0523abf2b7b7c5d#r59556136>
>

mrclay replied:
I thought that was assumed. To use crypto you need the DB anyway.

mrclay commented:
> Alternatively, we can substitute $session_id with another value, and use cookie value if session is not started.

👍 if you mean a new action token strategy based on, like a JWT in a cookie instead of session.

mrclay commented:
...but that's a can of worms to tackle by itself.

mrclay commented:
Have to also consider /refresh-token serves to keep the session alive.

mrclay commented:
I get logged out even when I'm not. Digging into it...

mrclay commented:
ElggSession::isLoggedIn() relies on `_elgg_session_boot` having been called to set the user entity. I think we can't *really* know if the user is logged in without a full boot that successfully loads the user entity, etc. 😢

mrclay commented:
`_elgg_session_boot()` would also make sure the user hasn't been banned, the persistent login cookie gets validated, etc. Open Q is whether `_elgg_services()->entityTable->get` can succeed without full boot?

hypeJunction commented:
Let's finally make a method to boot the app without all the init and ready
overhead.
On Apr 13, 2016 4:48 PM, "Steve Clay" <notifications@github.com> wrote:

> ElggSession::isLoggedIn() relies on _elgg_session_boot having been called
> to set the user entity. I think we can't *really* know if the user is
> logged in without a full boot that successfully loads the user entity, etc.
> 😢
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/Elgg/Elgg/pull/9675#issuecomment-209485267>
>

mrclay commented:
I feel we're making good progress wiring up services so they lazy-load what they need. I don't see much in `_elgg_session_boot` that requires a fully boot actually.

hypeJunction commented:
I will try to use session boot and see what it does. If it doesn't work, I will compromise on booting core and waiting till we have a lightdb or no plugin boot.

mrclay commented:
@hypeJunction hold up, PR incoming

mrclay committed:
chore(boot): allow partial boot

`Application::bootCore(true)` partially boots the system, not starting any
plugins or running any events.

mrclay commented:
See https://github.com/Elgg/Elgg/commit/fce677f9e044f0a3a25a4f04e566784c35c26de5

hypeJunction commented:
@mrclay This works quite well.

mrclay commented:
Low hanging fruit; cut around 30ms for me on a vanilla install.

hypeJunction commented:
You can't measure tests in ms ;)
On Apr 14, 2016 12:46 PM, "Steve Clay" <notifications@github.com> wrote:

> Low hanging fruit; cut around 30ms for me on a vanilla install.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/Elgg/Elgg/pull/9675#issuecomment-209875679>
>

mrclay commented:
- [ ] manually test calling bootCore() after bootCore(true)

hypeJunction commented:
@mrclay Should I just remove the deprecation and use the tested method from the `refresh_token` page handler? I don't want this to go to waste, but since the performance gain is so negligible, we can take advantage of the tests at least.

hypeJunction commented:
Moving to #9779
